src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSignature.cs

```diff
@@ -0,0 +1,254 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Security.Cryptography.Cose
+{
+    public sealed class CoseSignature
+    {
+        private readonly byte[] _encodedBodyProtectedHeaders;
+        internal readonly byte[] _encodedSignProtectedHeaders;
+        internal readonly byte[] _signature;
+
+        internal CoseSignature(CoseMultiSignMessage message, CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, byte[] encodedBodyProtectedHeaders, byte[] encodedSignProtectedHeaders, byte[] signature)
+            : this(protectedHeaders, unprotectedHeaders, encodedBodyProtectedHeaders, encodedSignProtectedHeaders, signature)
+        {
+            Message = message;
+        }
+
+        internal CoseSignature(CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, byte[] encodedBodyProtectedHeaders, byte[] encodedSignProtectedHeaders, byte[] signature)
+        {
+            ProtectedHeaders = protectedHeaders;
+            UnprotectedHeaders = unprotectedHeaders;
+            _encodedBodyProtectedHeaders = encodedBodyProtectedHeaders;
+            _encodedSignProtectedHeaders = encodedSignProtectedHeaders;
+            _signature = signature;
+        }
+
+        public CoseHeaderMap ProtectedHeaders { get; }
+        public CoseHeaderMap UnprotectedHeaders { get; }
+
+        private CoseMultiSignMessage? _message;
```

Comment:
The reason I separated the field from the property is because there's a moment where `_message` can be null, but it shouldn't when we access it from anywhere, hence, the `Debug.Assert(_message != null)` on `Message`'s getter.